tensor2tensor/data_generators/text_encoder.py

```diff
@@ -243,15 +254,22 @@ def _escaped_token_to_subtokens(self, escaped_token):
     """
     ret = []
     pos = 0
-    while pos < len(escaped_token):
-      end = len(escaped_token)
-      while True:
+    lesc = len(escaped_token)
+    while pos < lesc:
+      end = lesc
+      while end > pos:
         subtoken = self._subtoken_string_to_id.get(escaped_token[pos:end], -1)
         if subtoken != -1:
           break
         end -= 1
       ret.append(subtoken)
-      pos = end
+      if end > pos:
+        pos = end
+      else:
+        # This kinda should not happen, but it does. Cop out by skipping the
+        # nonexistent subtoken from the returned list.
+        # print("Unable to find subtoken in string '{0}'".format(escaped_token))
+        pos += 1
```

Comment:
If the `escaped_token[pos:end]` substring is for some reason not found in `self._subtoken_string_to_id`, the code decrements `end` and continues. In degenerate cases (which do appear in the test data sets, at least when I run `t2t-datagen` on Python3), `end` may become <= `pos`, meaning that the `escaped_token` slice becomes an empty string, which is never found in `self._subtoken_string_to_id`. In this case, the `while True` loop does not terminate and the program hangs.